### PR TITLE
Add support for rewriteTarget in file: upstreams

### LINKED ISSUE #2242: Regex path and rewriteTarget does not work for file:/// upstreams
The `rewriteTarget` option for HTTP/HTTPS upstreams allows for arbitrary rewriting of the request URI path that is passed to the upstream.  There are use cases where it would be useful to be able to make the same sort of transformations on a `file:///` upstream that serves a static file tree, but `rewriteTarget` does not work correctly with `file:` upstreams - every request produces a 404.

Consider a case where oauth2-proxy is protecting a single page application.  The whole application is contained in one `index.html` file plus a set of static assets for the JS and CSS:

- /var/www/app/
  - index.html
  - assets/
    - main.{hash}.js
    - style.{hash}.css

In order to support deep links into the application, we want to serve the assets folder as `https://example.com/assets/`, but _any other URL_ should return the content of `index.html`, so the SPA can inspect the page URL and route appropriately.  This would require (alpha) configuration like this:

```yaml
upstreamConfig:
  upstreams:
  # Send API calls to the backend
  - id: api
    path: ^/api/(.*)$
    rewriteTarget: /$1
    uri: http://backend.internal:8080/
  # Serve static assets
  - id: assets
    path: /assets/
    uri: file:///var/www/app/assets
  # For any other URL, return the index page
  - id: spa
    path: ^/(.*)
    rewriteTarget: /
    uri: file:///var/www/app
```

## Expected Behavior

- A request for `https://example.com/assets/main.12345.js` should return that file from the assets folder
- A request for `https://example.com/` should return the index page
- A request for `https://example.com/myprofile` should _also_ return the index page

## Current Behavior

- A request for `https://example.com/assets/main.12345.js` does return that file from the assets folder
- A request for `https://example.com/` gives "404 page not found"
- A request for `https://example.com/myprofile` gives "404 page not found"

## Possible Solution

At present, when we register a `file:///` upstream we use

```go
http.StripPrefix(path, http.FileServer(http.Dir(filesystemPath)))
```

where `path` is the `path:` option for the upstream and `filesystemPath` is the path component of the upstream `file:` URI.  But the `http.StripPrefix` handler always returns 404 if it is asked to handle a URL that does not have the configured prefix.  When the upstream `path` option is a regex it will likely never be an exact prefix of any actual URL path (because it will include metacharacters like `^`, `()`, `*`), so the `StripPrefix` handler will always 404.

The solution would be to only use the `StripPrefix` handler if the upstream is _not_ configured with a `rewriteTarget` - if a `rewriteTarget` is provided then the regex rewrite is what is responsible for mapping the matched path to the correct relative path within the target file tree.

## Your Environment

- Version used: 7.5.0, on Mac arm64 but this isn't a platform-specific problem

### PR DESCRIPTION
**Pull request by ianroberts**
Add support for regex `path` matching and `rewriteTarget` transformations when using a `file:` URL as the upstream.

## Description

Filesystem upstreams currently only support exact prefix matching, not the regex-based transformation mechanism using `rewriteTarget`.  This PR extends rewriting support to `file:` upstreams.  Essentially, when `rewriteTarget` is specified the result of the rewrite is used as the relative path from the directory referenced by the upstream's `uri` parameter.  Another way of thinking about it is that a plain (non-regex) file upstream

```yaml
upstreamConfig:
  upstreams:
  - id: files
    path: /static/
    uri: file:///var/www/staticfiles
```

is equivalent to the following with rewriting

```yaml
upstreamConfig:
  upstreams:
  - id: files
    path: ^/static/(.*)
    rewriteTarget: /$1
    uri: file:///var/www/staticfiles
```

Note that this is (and always has been!) different from the behaviour for non-regex `http(s)` upstreams:

```yaml
    path: /foo/
    uri: http://example.com/bar
```

means

```yaml
    path: ^/(foo/.*)
    rewriteTarget: /$1
    uri: http://example.com/bar
```

and sends `/foo/hello` to `http://example.com/bar/foo/hello`, whereas

```yaml
    path: /foo/
    uri: file:///bar
```

means

```yaml
    path: ^/foo/(.*)
    rewriteTarget: /$1
    uri: file:///bar
```

and sends `/foo/hello` to `file:///bar/hello` (not `/bar/foo/hello`).  I don't take a position on whether this difference is right or wrong but I've preserved backwards compatibility when `rewriteTarget` is not used.

## Motivation and Context

Fixes #2242  - as per the issue description, the use case I have in mind for this is using oauth2-proxy to serve files for a single-page application, where deep linking requires many different URLs to serve up the same single HTML page (the app JavaScript will inspect the page URL and route to the appropriate view of the app).

## How Has This Been Tested?

Tested locally with a variant of the config in the linked issue description.  I've added a test case to the `proxy_test.go` suite to verify the correct behaviour.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: pkg/upstream/file.go [hunk 1]
```diff
@@ -2,32 +2,66 @@ package upstream
...
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		reqURL, err := url.ParseRequestURI(req.RequestURI)
+		if err != nil {
+			http.Error(rw, "500 Internal Server Error", http.StatusInternalServerError)
+			return
+		}
+
+		req.URL = reqURL
+		handler.ServeHTTP(rw, req)
+	})
```
Comment (ianroberts) [inline pkg/upstream/file.go:64]: Question - is it better to fix this up here just for `file:` upstreams or would it make more sense to fix the rewrite middleware to adjust `req.URL` as well as `req.RequestURI`?

**Comment by ianroberts** [comment]
Rebased following 7.5.1 release

File: pkg/upstream/file.go (hunk 1 above)
Comment (JoelSpeed) [reply pkg/upstream/file.go:64]: Can you see if we use `req.URL` elsewhere? I'm wondering if past us left that alone so we had the original available _after_ the rewrite had happened

Comment (ianroberts) [reply pkg/upstream/file.go:64]: It looks like the main places that use `http.Request.URL` outside of tests are:

- `extractAllowedEntities` in `oauthproxy.go` uses `req.URL.Query()` but not the path
- `OAuthProxy.OAuthStart` uses the `Query()` from the `/oauth2/start` URL as a source for additional parameters to pass on to the provider's login URL (this was a previous PR of mine to allow for runtime overriding of things like `?prompt=consent`)
- `app/redirect/getters.go` uses `req.URL.RequestURI()` to determine where to redirect back to when a user gets auto-redirected to the login page when hitting a protected URL without a session - I assume this happens at an earlier stage in the handler chain, as the rewrite middleware only kicks in once we know they're allowed to see a page and we want to work out where to proxy to on the upstream
- the  CSRF cookie uses `req.URL.Query()` to extract the `state` parameter, which I think only applies during the OAuth flow, not on requests that will involve an upstream
- health check and readiness check middleware checks `req.URL.EscapedPath()` to see whether an incoming request is a health probe
- `redirectToHTTPS` builds a new URL from `req.URL.String()` and substitutes the scheme before redirecting the caller
- the request logger checks `url.Path` against a list of excluded paths - I guess this needs the original pre-rewrite path so where in the chain does the logger sit, is it before or after the rewrite middleware?
- `pkg/upstream/http.go` `setProxyDirector` copies `req.RequestURI` over to `req.URL.Opaque`

and the rewrite middleware itself of course.

Comment (ianroberts) [reply pkg/upstream/file.go:64]: @JoelSpeed I guess if we're in any way unsure then best to leave it the way I've done it here, only fixing up the `req.URL` specifically when it's about to go into the `http.FileServer` handler.

**Comment by ianroberts** [comment]
Rebased again

**Comment by ianroberts** [comment]
Not stale - just rebased again

**Comment by ianroberts** [comment]
@JoelSpeed is it possible to re-open this PR or do I need to submit a new one from the same source branch?

**Comment by ianroberts** [comment]
@kvanzuijlen re-created as #2589

**Pull request closed without merging**